### PR TITLE
feat(Farms): Warn users when their P is likely going out-of-range based on 7d data

### PR DESCRIPTION
**Pull request by Woody810**
### Summary
🚨📊🌐

This pull request adds a new feature to warn users when their v3 liquidity positions or farms are likely to go out of range based on historical price data. It modifies several components and files in the `pancake-frontend` app and the `uikit` package, such as `AprCalculator`, `RangeSelector`, `StepCounter`, and `RoiCalculator`. It also updates the `translations.json` file to add some new translation keys and values for the warning messages and labels.

> _Sing, O Muse, of the cunning code changes_
> _That the skilled developers wrought with their hands_
> _To enhance the `AprCalculator` and the `RoiCalculator`_
> _And to warn the users of the perilous price range._

### Walkthrough
*  Add warning sign and message to `AprCalculator` and `RoiCalculator` components when the user's chosen price range is likely to go out of range based on historical price data ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1R12), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1L54-R56), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1R92), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1R151-R179), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1L156-R193), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1L166-R200), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1R228-R231), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1L203-R242), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1R272), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-fda14e0d4e6e6151bd888953d6a631472fc742a475793fb2e13b847929fd9dc0R64), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-fda14e0d4e6e6151bd888953d6a631472fc742a475793fb2e13b847929fd9dc0R285), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-1da265a2b61cd6604fc701de6a520d682d0f800555bdd1e4c8aefb59af43ea8aR59-R67), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-1da265a2b61cd6604fc701de6a520d682d0f800555bdd1e4c8aefb59af43ea8aR110), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-1da265a2b61cd6604fc701de6a520d682d0f800555bdd1e4c8aefb59af43ea8aR327-R372), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-1da265a2b61cd6604fc701de6a520d682d0f800555bdd1e4c8aefb59af43ea8aR435), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-1da265a2b61cd6604fc701de6a520d682d0f800555bdd1e4c8aefb59af43ea8aR446))
*  Pass the `isWarnPriceOutOfRange` prop to the `RangeSelector`, `StepCounter`, and `RangeInputPanel` components and change the color of the step counter text to warning if the position is at risk of going out of range ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-14a297aa56e35bdc9f2b08afe5033407200cad3e7c44f0965602b57275b98f5aR21), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-14a297aa56e35bdc9f2b08afe5033407200cad3e7c44f0965602b57275b98f5aR35), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-14a297aa56e35bdc9f2b08afe5033407200cad3e7c44f0965602b57275b98f5aR60), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-14a297aa56e35bdc9f2b08afe5033407200cad3e7c44f0965602b57275b98f5aR75), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-49eb513f72eb32fe100a15fafd6a5efddf975336a21c1924c016033dd2d38d26R20), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-49eb513f72eb32fe100a15fafd6a5efddf975336a21c1924c016033dd2d38d26R35), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-49eb513f72eb32fe100a15fafd6a5efddf975336a21c1924c016033dd2d38d26R110), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-47bb12227d90c986843837ea996db5c39b7cbb1c85bc582de7747d28de2f5b09R28), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-47bb12227d90c986843837ea996db5c39b7cbb1c85bc582de7747d28de2f5b09R42), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-47bb12227d90c986843837ea996db5c39b7cbb1c85bc582de7747d28de2f5b09R68), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-47bb12227d90c986843837ea996db5c39b7cbb1c85bc582de7747d28de2f5b09R83), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-f2bc26f8a1831b5b9e66c06703097efc0ff53bf149a3e2f9ead5c1062393ac40R22), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-f2bc26f8a1831b5b9e66c06703097efc0ff53bf149a3e2f9ead5c1062393ac40R38), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-f2bc26f8a1831b5b9e66c06703097efc0ff53bf149a3e2f9ead5c1062393ac40R113))
*  Fetch the historical price data of the selected currency pair for the past 7 days using the `usePairTokensPrice` hook and the pool address derived from the `Pool` class ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-93f8dec4471df1307df92d3959703cfff58885d29c64ab692855fc06bf02f764L23-R26), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-93f8dec4471df1307df92d3959703cfff58885d29c64ab692855fc06bf02f764R52), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-93f8dec4471df1307df92d3959703cfff58885d29c64ab692855fc06bf02f764R415-R452), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-fda14e0d4e6e6151bd888953d6a631472fc742a475793fb2e13b847929fd9dc0R64))
*  Add fallback values of `''` to the `maxAmounts` expressions in the `onMax` and `onPercentInput` callbacks to prevent errors or crashes when calling the `toExact` or `multiply` methods on undefined or invalid values ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-93f8dec4471df1307df92d3959703cfff58885d29c64ab692855fc06bf02f764L431-R471), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-93f8dec4471df1307df92d3959703cfff58885d29c64ab692855fc06bf02f764L452-R492))
*  Add defensive checks to the `position` argument in the `onRoiCalculatorPositionInfo` callback to ensure that it has the required properties before accessing them ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1L156-R193), [link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-11372967482ecb262dc9c61cfcc4c28c7ac992e7a9bad352f6fce19f1940bdb1L166-R200))
*  Add a conditional style to the `Text` component in the `index.tsx` file based on the `isPositionBelowRange` or `isPositionAboveRange` states to change the color of the text to warning if the position is below or above the current market price ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-93f8dec4471df1307df92d3959703cfff58885d29c64ab692855fc06bf02f764L644-R685))
*  Add some new translation keys and values for the warning messages and labels related to the price range and the historical price data to the `translations.json` file ([link](https://github.com/pancakeswap/pancake-frontend/pull/6942/files?diff=unified&w=0#diff-eb2ab983e2cefc22516cb7814c86346f4b2bd2b971980952868173095d48f459L2489-R2502))

### CODE REVIEW

File: apps/web/src/views/AddLiquidityV3/formViews/V3FormView/index.tsx [hunk 1]
```diff
@@ -410,6 +412,44 @@ export default function V3FormView({
     [price, feeAmount, invertPrice, onBothRangeInput],
   )
 
+  const poolAddress = useMemo(() => pool && Pool.getAddress(pool.token0, pool.token1, pool.fee), [pool])
+  const prices7D = usePairTokensPrice(poolAddress, 1, baseCurrency?.chainId)
+  const priceKeyValues = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (Woody810) [reply apps/web/src/views/AddLiquidityV3/formViews/V3FormView/index.tsx]: fixed

File: apps/web/src/views/AddLiquidityV3/components/AprCalculator.tsx [hunk 2]
```diff
@@ -146,24 +148,56 @@ export function AprCalculator({
...
+  const priceKeyValues = useMemo(() => {
+    if (!prices7D || !prices7D.pairPriceData.length) {
+      return {
+        min: 0,
+        max: 0,
+        average: 0,
+      }
+    }
+
+    const _prices7D = prices7D.pairPriceData.map((p) => ({
```
[the remark this answers is not in the document]
Comment (Woody810) [reply apps/web/src/views/AddLiquidityV3/components/AprCalculator.tsx]: I'm not sure how to modify it here. Can you please explain it in more detail?🙏

File: packages/uikit/src/widgets/RoiCalculator/StepCounter.tsx [hunk 3]
```diff
@@ -108,6 +110,7 @@ export const StepCounter = memo(
               onUserInput={(val) => {
                 setLocalValue(val);
               }}
+              style={{ color: isWarnPriceOutOfRange ? "#FFB237" : "white" }}
```
Comment (chef-jojo) [inline packages/uikit/src/widgets/RoiCalculator/StepCounter.tsx:113]: use vars

File: packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx [hunk 4]
```diff
@@ -317,6 +328,25 @@ export function RoiCalculator({
     </Message>
   );
 
+  const tokenA = (currencyA ?? undefined)?.wrapped;
+  const tokenB = (currencyB ?? undefined)?.wrapped;
+  const isSorted = tokenA && tokenB && tokenA.sortsBefore(tokenB);
+  const leftPrice = Number(formatPrice(isSorted ? priceRange?.priceLower : priceRange?.priceUpper?.invert()));
+  const rightPrice = Number(formatPrice(isSorted ? priceRange?.priceUpper : priceRange?.priceLower?.invert()));
```
Comment (chef-jojo) [inline packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx:335]: the formatPrice can be '-' or '<0.0001'

File: packages/utils/usePairTokensPriceInverted.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+export const usePairTokensPriceInverted = (
+  prices?: {
+    pairPriceData: {
+      time: Date
+      value: number
+    }[]
+    maxPrice: number
+    minPrice: number
+    averagePrice: number
```
Comment (chef-jojo) [inline packages/utils/usePairTokensPriceInverted.ts:9]:
```
    maxPrice: number
    minPrice: number
    averagePrice: number
```
Any reason we cannot use this?

File: packages/utils/usePairTokensPriceInverted.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+export const usePairTokensPriceInverted = (
```
Comment (chef-jojo) [inline packages/utils/usePairTokensPriceInverted.ts:1]: avoid `use` for non hook function

File: packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx (hunk 4 above)
Comment (Woody810) [reply packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx:335]: It should be feasible to use the formatPrice function with formatFractions without returning '-' or '<0.0001'.

File: packages/utils/usePairTokensPriceInverted.ts (hunk 5 above)
Comment (Woody810) [reply packages/utils/usePairTokensPriceInverted.ts]: Since the `prices` will not be inverted.

**Pull request closed without merging**